src/schoology-extractor/schoology_extractor/mapping/attendance.py

```diff
@@ -0,0 +1,121 @@
+
+        for item in date_node["statuses"]["status"]:
+            if "attendances" not in item or "attendance" not in item["attendances"]:
+                continue
+
+            reshape = pd.DataFrame(
+                [
+                    {
+                        "enrollment_id": a["enrollment_id"],
+                        "EventDate": date_node["date"],
+                        "AttendanceStatus": a["status"],
+                    }
+                    for a in item["attendances"]["attendance"]
+                ]
+            )
+
+            df = df.append(reshape)
+
+    return df.convert_dtypes()
+
+
+def __get_status(status_code: int) -> str:
+    switcher = {1: "present", 2: "absent", 3: "late", 4: "excused"}
+    return switcher.get(status_code, f"Unknown status: {status_code}")
+
+
+def map_to_udm(attendance: list, section_associations: pd.DataFrame) -> pd.DataFrame:
+    """
+    Maps a DataFrame containing Schoology attendance events into the Ed-Fi LMS
+    Unified Data Model (UDM) format.
+
+    Parameters
+    ----------
+    attendance: list
+        List containing the API response from Schoology
+    section_associations: DataFrame
+        DataFrame containing UDM-mapped section associations
+
+    Returns
+    -------
+    DataFrame
+        A new DataFrame in the UDM attendance events format
+
+    Notes
+    -----
+    DataFrame columns are:
+        SourceSystemIdentifier: A unique number or alphanumeric code assigned to a user by the source system
+        SourceSystem: The system code or name providing the user data
+        Date: Attendance date, formatted as YYYY-mm-dd
+        AttendanceStatus: one of (present, absent, late, excused)
+        LMSUserSourceSystemIdentifier: source system identifier for the user
+        LMSUserLMSSectionAssociationSourceSystemIdentifier: source system identifier for the section association
+        EntityStatus: The status of the record
+        CreateDate: datetime at which the record was first retrieved
+        LastModifieDate: datetime when the record was modified, or when first retrieved
+    """
+    if len(attendance) == 0:
+        return pd.DataFrame()
+
+    df = _flatten_into_dataframe(attendance)
+
+    df["SourceSystem"] = constants.SOURCE_SYSTEM
+    df["EntityStatus"] = constants.ACTIVE
+    df["SourceSystemIdentifier"] = df.apply(
+        lambda row: f"{row['enrollment_id']}#{row['EventDate']}", axis=1
+    )
+
+    df["AttendanceStatus"] = df["AttendanceStatus"].apply(__get_status)
+
+    sa = section_associations[
+        [
+            "SourceSystemIdentifier",
+            "LMSUserSourceSystemIdentifier",
+            "LMSSectionSourceSystemIdentifier",
+        ]
+    ].copy()
+    # This data type conversion was required because Schoology is returning
+    # enrollment Id as an integer in the Attendance endpoint, but as a string
+    # with the Enrollment endpoint.
+    sa["SourceSystemIdentifier"] = sa["SourceSystemIdentifier"].apply(int)

```

Comment:
This drove me crazy. The merge statement below was working in unit testing, but failing in full application testing. Finally I realized that the failure was due to a type mismatch.

There is at least one more way to set a data type for a column: [astype()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.astype.html). They seem equally reasonable to me, and the approach above is the first one I learned so it was my default.

Reply:
I only wonder if astype() would avoid the iteration across all rows that the apply() requires.

Reply:
Since I'm going in to make a few other changes, I'm going to update this.

Reply:
Done